Start a chat with anyone

chiragsalian opened the pull request:
### Details
<Explanation of the change or anything fishy that is going on>

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/145241

### Tests
1. Create an account. Add a firstname, lastname and profile photo.
2. Open expensify cash and login with your expensify email.
3. Search for the account in step 1. Click the account and confirm the profile photo, first and names show up,
![image](https://user-images.githubusercontent.com/14356145/103041898-9a079d80-452c-11eb-8deb-e53204ae46ce.png)
4. Search for an email that does not exist in your account (say test@test123456.com), click the email to DM it and confirm you can send a message to the account. If you check the db you'll see its a closed account. This means that the user will get an email of the chat inviting them to use expensify cash.
5. Search for a phone number (say 5857527440), confirm +1 gets automatically prefixed. Click on the phone to DM it and confirm you can send a message to the account. Once again if you check the db you'll see its a closed account.
6. Type the phone number again, confirm you can edit and place whatever international code you want. Say we put +415857527440. click the phone number and confirm you see this error because our backend couldn't validate it.
![image](https://user-images.githubusercontent.com/14356145/103042125-64af7f80-452d-11eb-90ac-8c85b496ff64.png)
7. Type the email test@expensify.cash and click it. Confirm you see the error,
![image](https://user-images.githubusercontent.com/14356145/103042167-90cb0080-452d-11eb-8561-b1bde6e97712.png)

chiragsalian committed:
Initial code to chat with anybody

chiragsalian commented:
I have read the CLA Document and I hereby sign the CLA

chiragsalian committed:
using newUser method

chiragsalian committed:
feedback header

chiragsalian committed:
Fixes for group chat

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

chiragsalian committed:
Fixes to support inviting anyone

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

chiragsalian committed:
comments and corrections

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

chiragsalian committed:
handling errors

chiragsalian committed:
calling it what it is

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
-        });
+        let options = Array.from(matches);
+        if (options.length === 0) {
+            const searchStr = value;
```
Also maybe we should trim all white space at a minimum to take care of leading/trailing spaces?

chiragsalian replied:
Ah it doesn't need to be reassigned. I originally had `this.state.search` there and used searchStr to DRY it up but yeah I'll update to use value directly.

> Also maybe we should trim all white space at a minimum to take care of leading/trailing spaces?

Sure i can add that too.

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
+
+            const isPhoneNumber = /^[+|\d]+$/.test(searchStr);
+            if (searchStr && isPhoneNumber && Str.isValidPhone(searchStr)) {
+                const phoneNumber = searchStr.includes('+') ? searchStr : `+${this.props.countryCodeByIP}${searchStr}`;
```
I'm a little confused by all of this logic and finding it hard to tell if we are missing something. Just going to rubber duck you tell me if it sounds right (and then maybe we can add some comments).

- If the search string has a `+` we assume the user knows their country code and has included it
- Otherwise we are going to add the `+` and the country code
- But then the `login` (which I think we create the chat from) will either have the country code or not?

I think this will mean that if I type my number like this

`15165677123`

it will show up as

`+115165677123`

Which might not be expected... just wondering if there is anything more we can do here. But if not, we should explain this maybe so it's understood.

chiragsalian replied:
> If the search string has a + we assume the user knows their country code and has included it
Otherwise we are going to add the + and the country code
But then the login (which I think we create the chat from) will either have the country code or not?

All correct. 

> I think this will mean that if I type my number... Which might not be expected.

That's what tom asked me if we could add so i added it here. Sure i can add a comment above explaining this.

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
+            feedbackMessage = 'You\'ve reached the maximum number of participants for a group chat.';
+        } else if (this.state.search && this.state.options.length === 0) {
+            feedbackMessage = 'Don\'t see who you are looking for? Type their valid email/phone number to invite them.';
+        }
```
This works well but I wonder how you feel about creating a new UI component for the feedback message and then inlining everything? I think it looks a bit cleaner and encourages us to reuse the feedback message style and also clearly define the states that make the messages appear.

```js
const maxParticipantsReached = this.state.usersToStartGroupReportWith.length === MAX_GROUP_DM_LENGTH;
const hasSelectableOptions =  this.state.search && this.state.options.length > 0;
```

And then broke this out into 3 separate blocks using a shared UI component called `FeedbackMessage`

```js
{!hasSelectableOptions && (
    <FeedbackMessage
        message={'blah...'}
    />
)}
{maxParticipantsReached && (
    <FeedbackMessage
        message={'blah...'}
        title="blah"
    />
)}
{hasSelectableOptions && !maxParticipantsReached && (
    ....
)}
```

chiragsalian replied:
I'm neutral about creating another component just because I don't know where it can be reused at the moment and hence I'm unsure if it's preoptimization. Let me know if you feel otherwise and i can still create another component for FeedbackMessage. Although i'm unsure why `ChatSwitcherView` would call it 3 times rather than just once with the message it needs. I say that to prevent bugs, eg: say maxParticipantsReached is true and hasSelectableOptions is false. Then we could end up showing two messages which is not what we want. So i think the logic of creating the feedbackMessage would still remain and then we'd show the `FeedbackMessage` only if feedbackMessage has a value.

marcaaron replied:
Sure, I think if we used this pattern maybe one more time then we would want to create a component. Rule of three? And we have two already.

>Although i'm unsure why ChatSwitcherView would call it 3 times rather than just once with the message it needs

I just think it's easier to make sense out of what is happening. Set up your two conditions that you are looking for. Show what happens to the message for each condition with a clear name for each. Then no one has to parse the if/elseif or the extra variables + which one has a title and which one does... well above the context of where they are used. The alternative is nice and clean and tells a story.

It's fine if we don't want to do this now, but I think what you have is hard to read so that's the suggestion I had.

chiragsalian replied:
> Sure, I think if we used this pattern maybe one more time then we would want to create a component. Rule of three?

Sounds good with me. Curious to pick your brain on this to understand how you think. Here is my usual thought pattern and you can let me know if there is something incorrect. I usually think of breaking something in a component if it can be reused in one other place. I don't really think of breaking it into another component if it's just used once in a single component (even if it can be called many times in one component).
Because,
```
{condition1 && (
    <FeedbackMessage
        message={'blah...'}
    />
)}
{condition2 && (
    <FeedbackMessage
        message={'blah...'}
        title="blah"
    />
)}
{condition3 && (
    <FeedbackMessage
        message={'blah...'}
        title="blah"
    />
)}
{!condition1 && !condition2 && !condition3 && (
    ....
)}
```
Still means that FeedbackMessage if shown is just shown once and then in that case it feels like the code would be easier to read like so,
```
let feedBackMessage = '';
if (condition1) {
    feedBackMessage = 'message1';
} else if (condition1) {
    feedBackMessage = 'message2';
} else if (condition3) {
    feedBackMessage = 'message3';
}

{feedBackMessage && (
    <FeedbackMessage
        message={feedBackMessage}
        title="blah"
    />
)}
{!feedBackMessage && (
    ....
)}
``` 

> Then no one has to parse the if/elseif or the extra variables + which one has a title and which one does

Is it generally a recommended pattern to keep lesser logic like this and instead have more variables like hasSelectableOptions and maxParticipantsReached and then have the render handle them accordingly rather than thinking about the least number of component calls? I was just curious if there is some general pattern i should improve on wrt react.

marcaaron replied:
I don't mind having extra variables set up if it helps describe a state later on. It's not really a React thing to me just a general readability thing and can help with refactoring or separating concerns in a complicated component like this.

I am also in the process of refactoring this anyway so I wont block on this. One thing that will happen eventually is the logic for these two "feedback messages" will not exist in the same place anymore. This will eventually just be a prop of "feedbackMessage" and "feedbackTitle". Breaking the if/else into two separate concerns will make it easier to see when the two messages show. One is only relevant to group chats the other is relevant to both group chats and new chats.

marcaaron replied:
Also, I think this kind of small helper text style will end up being used a lot. That's the only reason I think a component would be good here. But we cant wait on that.

marcaaron commented on src/styles/styles.js:
```diff
@@ -877,10 +877,6 @@ const styles = {
-    chatSwitcherMessage: {
-        padding: 20,
-    },
-
```
👍

marcaaron reviewed (changes requested):
Testing well but wondering if we want to get rid of the `@expensify.sms` and also had a few other comments and suggestions.

marcaaron commented:
![Screen Shot 2020-12-24 at 12 36 34 PM](https://user-images.githubusercontent.com/32969087/103111025-815ebc80-45ec-11eb-8d0e-b770be551b66.png)

`@expensify.sms` is visible. I think it's worth fixing because QA will spot it immediately

It's interesting to see that when we add a phone number with country code but no `+` we see a `+[country code]` added to the number with the country code.

![Screen Shot 2020-12-24 at 1 16 07 PM](https://user-images.githubusercontent.com/32969087/103111034-9d625e00-45ec-11eb-84c8-f7c7a13bcf99.png)

Is this something we have plans to address?

chiragsalian commented on src/libs/actions/GeoLocationUtils.js:
```diff
@@ -0,0 +1,15 @@
+        });
+}
+
+export default fetchCountryCodeByRequestIP;
```
Sure, is that usually prefferred?

chiragsalian replied:
I made the changes here but it feels a little weird. Let me know if there is a reason we'd prefer the change or not.

marcaaron replied:
I like it for consistency. I think whatever our action files export all other action files should export so we never have to think about it. Theres no set rule on this yet AFAIK, but we can chat about it sometime and decide.

chiragsalian commented on src/ONYXKEYS.js:
```diff
@@ -30,6 +30,10 @@ export default {
 
+    // Saves the current country code by IP which is displayed when the user types a phone number without
+    // international code
+    COUNTRY_CODE_BY_IP: 'countryCodeByIP',
```
Sure, i soft disagee. Atm i named it what it [is](https://github.com/Expensify/Web-Expensify/blob/2d17b9582b720e2480e4ff6fe59d3802f1aa17e3/lib/Geo/LocationUtils.php#L52). 

> Maybe one day the user is able to set their country

Planning for that sounds like preoptimization and when that time comes then we can change the const. Thoughts?

Also as i write this out is there a way to clear unused ONYX saved data from the users local storage?

marcaaron replied:
Well if you are soft on this can you please change it? If we can name something correctly and consistently on the first pass I say let's do it. All of the other keys in this file describe the data they do not tell us anything about where it came from. If for no other reason do it for consistency and so the next person who saves something doesn't follow this example.

chiragsalian commented on src/libs/actions/Report.js:
```diff
@@ -442,6 +442,10 @@ function fetchOrCreateChatReport(participants) {
         .then((data) => {
+            if (data.jsonCode !== 200) {
+                throw new Error(data.message);
+            }
```
So if I alert here and do a return it will fall into the next `.then` and then i'd have to handle it there as well hence duplicating some code. By doing a throw and a catch it keeps the flow clean. 

> Maybe something else causes an error here and you are now alerting whatever it is in the catch()

Not sure why this is a bad thing. If we see an interesting alert of a case we don't want to handle or catch differently then we update the flow accordingly when that happens.

marcaaron replied:
>it will fall into the next .then 

Not too sure what you mean as there is no "next `.then()`"?

>Not sure why this is a bad thing

It's bad because you are alerting to the user a message that is not customer facing and could be ... anything?

chiragsalian replied:
> Not too sure what you mean as there is no "next .then()"?

So if i replace,
```
throw new Error(data.message);
```
with,
```
alert(data.message);
return;
```
then it will fall into this ["then"](https://github.com/Expensify/Expensify.cash/blob/e05454f16a2d8d01f9b513bfa81457739de225f5/src/libs/actions/Report.js#L461-L481) block and throw an error because `data.reports[reportID]` does not exist. So then I've to have an early return there too when data is not what we expect. Hence doing a throw from the first "then" block and then using a catch felt clean. 

> It's bad because you are alerting to the user a message that is not customer facing and could be ... anything?

Well i guess that's the issue with this first phase solution. Either i throw an alert for any flow or handle the error redundantly in those [two "then" conditions](https://github.com/Expensify/Expensify.cash/blob/e05454f16a2d8d01f9b513bfa81457739de225f5/src/libs/actions/Report.js#L444-L481). So yeah let me know your thoughts. One solution i can think off is to save a list of accepted alert messages in CONST then have the catch error check if its part of those to display it. If its not then it would do a console error log instead. How does that sound?

marcaaron replied:
I like the idea about checking for a specific constant error message. But I think it would be easier to just alert when you do not have the necessary result. If there is no `data` in the second `.then()` then return early again and that will solve the problem.

chiragsalian replied:
Alright, updated this to alert when the first "then" result is not what i expect and I've updated the next "then" function to return early when there are no reports.

chiragsalian commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
-        this.setState({
-            options: Array.from(matches),
-        });
+        let options = Array.from(matches);
```
That logic would also work but i don't see any benefit over what i currently have. I think with the let and assignment its even more obvious there is just one option added from the next block. Is there a reason you would prefer const here?

marcaaron replied:
Maybe personal bias/instinct, but if something doesn't really need to be reassigned and can be a `const` then we should prefer a `const` so you don't have to spend time thinking about which one would be better. However, this is not in any of of style guides so I wont push too hard on it.

chiragsalian commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
+        let options = Array.from(matches);
+        if (options.length === 0) {
+            const searchStr = value;
+            if (searchStr && searchStr.includes('@') && Str.isValidEmail(searchStr)) {
```
Ah yes, i added `Str.isValidEmail` later. I'll remove the check for `@`

chiragsalian commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,47 @@ class ChatSwitcherView extends React.Component {
+                }];
+            }
+
+            const isPhoneNumber = /^[+|\d]+$/.test(searchStr);
```
Yup, i added `Str.isValidPhone` later. I'll remove `isPhoneNumber`

chiragsalian commented:
@expensify.sms is visible. Ah yeah good catch. I totally didn't think about removing that. Shall update.

chiragsalian committed:
cleanup, icon fixes and removing phone number domain

chiragsalian committed:
removing unneeded Str

chiragsalian commented:
Addressed review comments, Removed `@expensify.sms` when we display phone numbers and corrected the code so that external domains also show up with default avatars.
![image](https://user-images.githubusercontent.com/14356145/103141731-60aa6b80-46ad-11eb-9c90-763a903a7783.png)

chiragsalian committed:
variable and flow cleanup

chiragsalian committed:
Merge branch 'master' into chirag-invite-anybody

marcaaron commented on src/CONST.js:
```diff
@@ -5,15 +5,18 @@ const CONST = {
         },
     },
+    LOGIN: {
+        EXPENSIFY_SMS_SUFFIX: '@expensify.sms'
```
This should exist in `expensify-common`

https://github.com/Expensify/expensify-common/blob/d1bfacca6fb92f7e89a2a9cdf0863fad6481c0f2/lib/CONST.jsx#L488

chiragsalian commented:
Updated code to address the remaining comments.

marcaaron commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -57,6 +57,12 @@ function getAvatar(personalDetail, login) {
 
+    // If we have a number like +15857527441@expensify.sms then let's remove @expensify.sms
+    // so that the option looks cleaner in our UI.
+    if (login.includes(CONST.LOGIN.EXPENSIFY_SMS_SUFFIX)) {
```
This is probably all you need -> https://github.com/Expensify/expensify-common/blob/d1bfacca6fb92f7e89a2a9cdf0863fad6481c0f2/lib/str.js#L909-L918

chiragsalian replied:
Oh i missed that Str library probably has checks for these. I think I'll need both Str isSMSLogin and removeSMSDomain so shall update.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -442,6 +442,10 @@ function fetchOrCreateChatReport(participants) {
 
         .then((data) => {
+            if (data.jsonCode !== 200) {
+                alert(data.message);
```
Should we `return`? We are still making a call to `Get` `reportStuff` with no reportID ?

chiragsalian replied:
Yes, my mistake.

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,39 @@ class ChatSwitcherView extends React.Component {
-        });
+        const options = Array.from(matches);
+        if (options.length === 0) {
+            if (value && (Str.isValidEmail(value) || Str.isValidPhone(value))) {
```
These two if statements can be combined

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -461,12 +466,39 @@ class ChatSwitcherView extends React.Component {
+        if (options.length === 0) {
+            if (value && (Str.isValidEmail(value) || Str.isValidPhone(value))) {
+                let login = value;
+                if (Str.isValidPhone(value)) {
```
NAB this could me made into a variable `isPhone`

chiragsalian replied:
I refrained from making this a variable since i don't want to call `Str.isValidPhone` unless options.length === 0 and value is not empty. So if i did,
```
isPhone = options.length === 0 && value && Str.isValidPhone(value);
```
it feels a little weird so I'm just calling `Str.isValidPhone` for now. I didn't want to do just,
```
isPhone = Str.isValidPhone(value);
```
before the if condition because i think keytyping it pretty slow and it feels like string checks on keytype would slow it down further hence i wanted to string check only when options.length === 0 and value is not empty. Leaving it as is for now but if you feel differently let me know.

marcaaron commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -116,7 +116,7 @@ const SidebarLinks = (props) => {
                                 alternateText: participantDetails ? participantDetails.login : '',
                                 type: participantDetails ? 'user' : 'report',
-                                icon: participantDetails ? participantDetails.avatarURL : '',
+                                icon: report.icon,
```
NAB, could we maybe leave a comment here to explain why the icon is on the report? I think this is being worked on by some contractors and it's not super obvious. Something like 

>The icon for the row is set when the fetching personal details via `PersonalDetails.getFromReportParticipants()`

marcaaron reviewed (changes requested):
Looks good. Just a few small comments.

chiragsalian committed:
Using Str removeSMSDomain and fixing a flow

chiragsalian commented:
Updated. Fixed a case where the alternate text showed up with expensify.sms but now it doesn't.
![image](https://user-images.githubusercontent.com/14356145/103255397-096efa00-493e-11eb-9e29-d6c1a60673df.png)

marcaaron commented:
Taking this off HOLD now that https://github.com/Expensify/Web-Expensify/pull/29942 is on production.

marcaaron reviewed (approved):
LGTM

chiragsalian commented:
Awesome, thanks for taking it off hold and merging it 🙇
